extensions/INTEL/SPV_INTEL_masked_gather_scatter.asciidoc

```diff
@@ -0,0 +1,290 @@
+_physical pointer type_ is allowed for *OpVariable* it is implicitly possible to be used by *OpStore* and *OpLoad* which can
+store/load through a pointer to this vector.
+
+Change the _Overview_ of *OpVariable* as follows:
+Allocate an object or a vector of objects in memory, resulting in a pointer or appropriately a vector of pointers to it,
+which can be used with *OpLoad* and *OpStore*.
+Change the _Result Type_ of *OpVariable* as follows:
+_Result Type_ must be an *OpTypePointer* or a _vector_ with _physical pointer type_ _Component Type_.
+Its _Type_ operand is the type of object or vector of objects in memory.
+
+Modify *OpAccessChain* (implicitly modifies *OpInBoundsAccessChain*, *OpPtrAccessChain* and *OpInBoundsPtrAccessChain* instructions)
+Change the _Base_ as follows:
+_Base_ must be a pointer, pointing to the base of a composite object or a _vector_ of _physical pointer type_.
+
+Allow _vector_ of _physical pointer type_ to be the type of _Operand 1_ and _Operand 2_ of *OpPtrEqual*, *OpPtrNotEqual* and
+*OpPtrDiff* instructions. If operands are vectors of pointers, then the _Result Type_ of *OpPtrEqual* and *OpPtrNotEqual* is a
+vector with boolean _Component Type_ and _Result Type_ of *OpPtrDiff* is a vector with integer _Component Type_.
+
+Add the following new entries:
+
+[cols="1,1,6*3",width="100%"]
+|=====
+7+|[[OpMaskedGatherINTEL]]*OpMaskedGatherINTEL* +
+ +
+Reads values from a vector of pointers gathering them into one vector. Returns the gathered vector. Memory access
+is specified by a mask instruction parameter. +
+ +
+'Result Type' is a type of the gathered vector. Its _Component Type_ must be the same as the base type of
+'PtrVector'.
+ +
+'PtrVector' is a _vector_ with _physical pointer type_ _Component Type_, containing addresses from where the instruction reads. +
+ +
+'Alignment' is an unsigned 32-bit integer literal whose value is
+either 0 or a power of two. When the value is not 0, it is an assertion that
+each pointer value in 'PtrVector' has this alignment. The behavior is undefined if
+any pointer value in 'PtrVector' does not have this alignment. +
+ +
+'Mask' is a vector of boolean values with the same number of elements as the _Result Type_. It specifies which elements of
+'PtrVector' should be gathered. +
+ +
+'FillEmpty' is used to fill the masked-off lanes of the result. It must be of the same type as the _Component Type_ of _Result Type_. +
+
+1+|Capability: +
+*{capability_name}*
+1+| 7 | {OpMaskedGatherINTEL_token}
+| '<id>' +
+'Result Type'
+|'Result <id>'
+| '<id>' +
+'PtrVector'
+| '<literal>' +
+'Alignment'
+| '<id>' +
+'Mask'
+| '<id>' +
+'FillEmpty'
+|=====
+
+[cols="1,1,4*3",width="100%"]
+|=====
+5+|[[OpMaskedScatterINTEL]]*OpMaskedScatterINTEL* +
+ +
+Writes values from a vector to the corresponding memory address of the given vector of pointers. Memory access
+is specified by a mask instruction parameter. +
+ +
+'InputVector' is a _vector_ of values to scatter. +
```

Comment:
Not a blocker, but OpStore puts the pointer operand first. Not sure if you want to follow that precedent.

Reply:
Good comment. I agree, that the order should follow OpStore. The problem though is that we have implemented the instruction across the software stack and changing order now would be backward incompatible. I see two possible ways to deal with it:
1. Change Op code and the order now, leaving for us to support both old and new Op codes and orders;
2. If somebody express an interest to make this extension EXT/KHR or add the instructions to the core spec - we can change the Op code and the order then.

I'd slightly prefer to stick with the option 2, wdyt?

Reply:
If you've implemented it already I agree it is not worth changing now. We can think about option 2 if there is ever wider adoption.